keps/sig-auth/2579-pod-isolation-policy/README.md

```diff
@@ -0,0 +1,951 @@
+     - Doesn’t compose well
+     - Mutation priority can be unexpected
+
+However, we still feel that Kubernetes should include a mechanism to prevent privilege escalation
+through the create-pod permission.
+
+### Goals
+
+Replace PodSecurityPolicy without compromising the ability for Kubernetes to limit privilege
+escalation out of the box. Specifically, create/update pod permission should not be equivalent
+to root-on-node (or cluster).
+
+#### Requirements
+
+1. Validating only (i.e. no changing pods to make them comply with policy)
+2. Safe to enable in new AND upgraded clusters
+    - Dryrun policy changes and/or Audit-only mode
+3. Built-in in-tree controller
+4. Capable of supporting Windows in the future, if not in the initial release
+    - Don’t automatically break windows pods
+5. Must be responsive to Pod API evolution across versions
+6. (fuzzy) Easy to use, don’t need to be a kubernetes/security/linux expert to meet the basic objective
+
+Nice to have:
+
+1. Exceptions or policy bindings by requesting user
+2. Extensible: should work with custom policy implementations without whole-sale replacement
+3. Windows support in the initial release
+4. Enabled by default (not in alpha)
+5. Provide an easy migration path from PodSecurityPolicy
+6. Enforcement on pod-controller resources (i.e. things embedding a PodTemplate)
+
+### Non-Goals
+
+1. Provide a configurable policy mechanism that meets the needs of all use-cases
+2. Limit privilege escalation and other attacks beyond the pod to host boundary
+   (e.g. policies on services, secrets, etc.)
+
+
+## Proposal
+
+The three profile levels (privileged, baseline, restricted) of the [Pod Security Standards] will
+be hardcoded into the new admission plugin. Changes to the standards will be tied to the Kubernetes
+version that they were introduced in.
+
+### API
+
+<<[UNRESOLVED]>>
+
+The annotation prefix `podisolationpolicy.kubernetes.io` is open to discussion. The proposed
+prefix structure is `<featurename>.kubernetes.io`, and depends on the final name we settle on.
+
+<<[/UNRESOLVED]>>
+
+Policy application is controlled based on labels on the namespace. The following labels are supported:
+```
+podisolationpolicy.kubernetes.io/allow: <policy level>
+podisolationpolicy.kubernetes.io/allow-version: <policy version>
+podisolationpolicy.kubernetes.io/audit: <policy level>
+podisolationpolicy.kubernetes.io/audit-version: <policy version>
+podisolationpolicy.kubernetes.io/warn: <policy level>
+podisolationpolicy.kubernetes.io/warn-version: <policy version>
+```
+
+**Allow:** Pods meeting the requirements of the allowed level are allowed. Violations are rejected
+in admission.
+
+**Audit:** Pods and [templated pods] meeting the requirements of the audit
+policy level are ignored. Violations are recorded in a
+`podisolationpolicy.kubernetes.io/audit: <violation>` annotation on the audit event for the request.
```

Comment:
will we also place this annotation on the pods themselves so that users have a way of knowing them?

Reply:
@liggitt and I briefly discussed this, but dismissed it as it violates the requirement to be non-mutating. I think this use case is mostly covered by dry-running policy changes?

Reply:
one reason for giving independent control over admin-facing (audit) and user-facing (warn) actions was to give admins flexibility to audit things without concerning their users, or warn users without muddying audit logs. If you want both admin-facing and user-facing actions, set both audit and warn to the same level?

Reply:
oops, meant to leave that comment below on the "are warnings tracked in the audit log automatically" comment